File: src/core/pipeline.ts
import { randomUUID } from "node:crypto";
import type { ProviderCircuitBreaker } from "../strategies/circuit-breaker.js";
import type { IdempotencyResolver } from "../strategies/idempotency.js";
import type { RateLimiter } from "../strategies/rate-limit.js";
import type { RetryStrategy } from "../strategies/retry.js";
import { sanitizeMeridianError } from "./error-sanitizer.js";
import { sanitizeMetric, sanitizeObject } from "./observability-sanitizer.js";
import { sanitizeRequestOptions } from "./request-sanitizer.js";
import type {
  ErrorContext,
  NormalizedResponse,
  ObservabilityAdapter,
  Policy,
  PolicyContext,
  ProviderAdapter,
  RawResponse,
  RequestContext,
  RequestOptions,
  ResponseContext,
} from "./types.js";
import { MeridianError } from "./types.js";

import type { AuthConfig, AuthToken } from "./types.js";

export interface PipelineConfig {
  provider: string;
  adapter: ProviderAdapter;
  authConfig: AuthConfig;
  circuitBreaker: ProviderCircuitBreaker;
  rateLimiter: RateLimiter;
  retryStrategy: RetryStrategy;
  idempotencyResolver: IdempotencyResolver;
  observability: ObservabilityAdapter[];
  timeout: number | undefined;
  autoGenerateIdempotencyKeys?: boolean;
  sanitizerOptions?:
    | {
        redactedKeys?: string[];
        piiRedaction?: boolean | undefined;
        indiaMode?: boolean | undefined;
      }
    | undefined;
  compliance?:
    | {
        piiRedaction?: boolean | undefined;
        auditLog?: boolean | undefined;
        indiaMode?: boolean | undefined;
      }
    | undefined;
  onRawRequest?: (
    requestId: string,
    endpoint: string,
    method: string,
    options: RequestOptions,
  ) => void;
  policies?: Policy[];
}

export class RequestPipeline {
  private config: PipelineConfig;
  private cachedToken: AuthToken | null = null;

  constructor(config: PipelineConfig) {
    this.config = config;
  }

  private async getAuthToken(): Promise<AuthToken> {
    if (this.cachedToken) {
      if (this.cachedToken.expiresAt) {
        const bufferMs = 60_000;
        if (this.cachedToken.expiresAt.getTime() - bufferMs > Date.now()) {
          return this.cachedToken;
        }
        // Token expired or about to expire — fall through to re-fetch
      } else {
        // No expiry = static API key — cache forever
        return this.cachedToken;
      }
    }
    const token = await this.config.adapter.authStrategy(this.config.authConfig);
    this.cachedToken = token;
    return token;
  }

  private safelyBroadcastObservability(
    action: (adapter: ObservabilityAdapter) => void,
    actionName: string,
  ): void {
    const errors: Array<{ adapter: string; error: unknown }> = [];

    for (const obs of this.config.observability) {
      try {
        action(obs);
      } catch (error) {
        errors.push({
          adapter: obs.constructor?.name || "UnknownObservabilityAdapter",
          error,
        });
      }
    }

    if (errors.length > 0) {
      const errorSummary = errors
        .map(
          ({ adapter, error }) =>
            `  - ${adapter}: ${error instanceof Error ? error.message : String(error)}`,
        )
        .join("\n");

      console.error(
        `[Meridian] Observability failure in ${actionName} (${errors.length}/${this.config.observability.length} adapters failed):\n${errorSummary}`,
      );
    }
  }

  async execute<T>(endpoint: string, options: RequestOptions = {}): Promise<NormalizedResponse<T>> {
    const requestId = randomUUID();
    const method = options.method ?? "GET";
    const startTime = Date.now();

    if (this.config.autoGenerateIdempotencyKeys && !options.idempotencyKey) {
      options.idempotencyKey = randomUUID();
    }

    const idempotencyLevel = this.config.idempotencyResolver.getIdempotencyLevel(
      method,
      endpoint,
      options,
    );

    const identity = options.identity;

    this.config.onRawRequest?.(requestId, endpoint, method, options);

    if (this.config.policies?.length) {
      let ctx: PolicyContext = { provider: this.config.provider, endpoint, method };
      if (options.body !== undefined) (ctx as { body?: unknown }).body = options.body;
      if (options.headers !== undefined)
        (ctx as { headers?: Record<string, string> }).headers = options.headers;
      if (options.query !== undefined)
        (ctx as { query?: Record<string, string | number | boolean> }).query = options.query;
      for (const policy of this.config.policies) {
        const decision = policy.evaluate(ctx);
        if (!decision.allow) {
          throw new MeridianError(
            `Policy "${policy.name}" blocked request: ${decision.reason}`,
            "validation",
            this.config.provider,
            false,
            requestId,
          );
        }
        if (decision.transform !== undefined) {
          const patch = decision.transform(ctx);
          const updated = { ...options };
          if (patch.body !== undefined) {
            updated.body = patch.body;
            ctx = { ...ctx, body: patch.body };
          }
          if (patch.query !== undefined) {
            updated.query = patch.query;
            ctx = { ...ctx, query: patch.query };
          }
          if (patch.headers !== undefined) {
            updated.headers = patch.headers;
            ctx = { ...ctx, headers: patch.headers };
          }
          Object.assign(options, updated);
        }
      }
    }

    const sanitizerOpts = {
      ...this.config.sanitizerOptions,
      piiRedaction:
        this.config.compliance?.piiRedaction || this.config.sanitizerOptions?.piiRedaction,
      indiaMode: this.config.compliance?.indiaMode || this.config.sanitizerOptions?.indiaMode,
    };

    const sanitizedOptions = sanitizeRequestOptions(options, sanitizerOpts);

    const requestContext: RequestContext = {
      provider: this.config.provider,
      endpoint,
      method,
      requestId,
      timestamp: new Date(),
      options: sanitizedOptions,
      identity,
    };

    this.safelyBroadcastObservability((obs) => obs.logRequest(requestContext), "logRequest");

    try {
      await this.getAuthToken(); // warm the cache before acquiring the rate limiter

      await this.config.rateLimiter.acquire();

      let retryCount = 0;
      const response = await this.config.retryStrategy.execute(
        async () => {
          retryCount++;
          return await this.config.circuitBreaker.execute(async () => {
            try {
              return await this.executeHttpRequest(endpoint, options, await this.getAuthToken());
            } catch (err) {
              // On 401, attempt a one-shot token refresh if the adapter supports it
              if (
                typeof err === "object" &&
                err !== null &&
                "status" in err &&
                (err as { status: number }).status === 401 &&
                typeof (this.config.adapter as { refreshAuth?: unknown }).refreshAuth === "function"
              ) {
                const currentToken = this.cachedToken;
                if (currentToken) {
                  const refreshed = await (
                    this.config.adapter as Required<Pick<ProviderAdapter, "refreshAuth">>
                  ).refreshAuth(this.config.authConfig, currentToken);
                  this.cachedToken = refreshed;
                  return await this.executeHttpRequest(endpoint, options, refreshed);
                }
              }
              throw err;
            }
          });
        },
        idempotencyLevel,
        !!options.idempotencyKey,
      );

      const rateLimitInfo = this.config.adapter.rateLimitPolicy(response.headers);
      this.config.rateLimiter.updateFromHeaders(response.headers, rateLimitInfo);

      const normalized = this.config.adapter.parseResponse(response);

      normalized.meta.requestId = requestId;

      const duration = Date.now() - startTime;

      normalized.meta.trace = {
        retries: retryCount - 1,
        latency: duration,
        circuitBreaker: this.config.circuitBreaker.getStatus().state,
        rateLimitRemaining: rateLimitInfo.remaining,
      };

      const responseContext: ResponseContext = {
        provider: this.config.provider,
        endpoint,
        method,
        requestId,
        statusCode: response.status,
        duration,
        timestamp: new Date(),
        identity,
        trace: normalized.meta.trace,
      };

      this.safelyBroadcastObservability((obs) => obs.logResponse(responseContext), "logResponse");

      this.safelyBroadcastObservability(
        (obs) =>
          obs.recordMetric(
            sanitizeMetric(
              {
                name: "meridian.request.count",
                value: 1,
                tags: {
                  provider: this.config.provider,
                  endpoint,
                  status: String(response.status),
                },
                timestamp: new Date(),
              },
              this.config.sanitizerOptions,
            ),
          ),
        "recordMetric:request.count",
      );

      this.safelyBroadcastObservability(
        (obs) =>
          obs.recordMetric(
            sanitizeMetric(
              {
                name: "meridian.request.duration",
                value: duration,
                tags: {
                  provider: this.config.provider,
                  endpoint,
                },
                timestamp: new Date(),
              },
              this.config.sanitizerOptions,
            ),
          ),
        "recordMetric:request.duration",
      );

      return normalized as NormalizedResponse<T>;
    } catch (error) {
      const duration = Date.now() - startTime;

      let meridianError: MeridianError;
      try {
        const adapterError = this.config.adapter.parseError(error);

        meridianError = sanitizeMeridianError(adapterError, this.config.provider, requestId);
      } catch (parseError) {
        meridianError = sanitizeMeridianError(
          {
            message: error instanceof Error ? error.message : String(error),
            metadata: {
              originalError: error instanceof Error ? error.message : error,
              parseError: parseError instanceof Error ? parseError.message : parseError,
            },
          },
          this.config.provider,
          requestId,
        );
      }

      if (meridianError.category === "rate_limit" && meridianError.retryAfter) {
        this.config.rateLimiter.handle429(
          Math.floor((meridianError.retryAfter.getTime() - Date.now()) / 1000),
        );
      }

      let errorContext: ErrorContext = {
        provider: this.config.provider,
        endpoint,
        method,
        requestId,
        error: meridianError,
        duration,
        timestamp: new Date(),
        identity,
      };

      try {
        const sanitizedError = { ...meridianError } as any;
        if (sanitizedError.metadata) {
          sanitizedError.metadata = sanitizeObject(
            sanitizedError.metadata,
            this.config.sanitizerOptions,
          ) as Record<string, unknown>;
        }
        errorContext = { ...errorContext, error: sanitizedError };
      } catch {}

      this.safelyBroadcastObservability((obs) => obs.logError(errorContext), "logError");

      this.safelyBroadcastObservability(
        (obs) =>
          obs.recordMetric(
            sanitizeMetric(
              {
                name: "meridian.request.error",
                value: 1,
                tags: {
                  provider: this.config.provider,
                  endpoint,
                  errorCategory: meridianError.category,
                },
                timestamp: new Date(),
              },
              this.config.sanitizerOptions,
            ),
          ),
        "recordMetric:request.error",
      );

      throw meridianError;
    }
  }

  private async executeHttpRequest(
    endpoint: string,
    options: RequestOptions,
    authToken: AuthToken,
  ): Promise<RawResponse> {
    const timeout = this.config.timeout ?? 30000;

    const builtRequest = this.config.adapter.buildRequest({
      endpoint,
      options,
      authToken,
    });

    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), timeout);

    const fetchOptions: RequestInit = {
      method: builtRequest.method,
      headers: builtRequest.headers,
      signal: controller.signal,
    };
    if (builtRequest.body !== undefined) {
      fetchOptions.body = builtRequest.body;
    }

    try {
      const response = await fetch(builtRequest.url, fetchOptions);

      let body: unknown;
      try {
        const contentType = response.headers.get("content-type");
        if (contentType?.includes("application/json")) {
          body = await response.json();
        } else {
          body = await response.text();
        }
      } catch {
        body = {};
      }

      const headersMap = new Headers();
      response.headers.forEach((value, key) => {
        headersMap.set(key, value);
      });

      if (!response.ok) {
        throw {
          status: response.status,
          headers: headersMap,
          body,
        };
      }

      return {
        status: response.status,
        headers: headersMap,
        body,
      } as RawResponse;
    } catch (error) {
      if (error instanceof Error && error.name === "AbortError") {
        throw new MeridianError(
          `Request timeout after ${timeout}ms`,
          "network",
          this.config.provider,
          true,
          "",
          {
            timeout,
            url: builtRequest.url,
            method: builtRequest.method,
          },
          undefined,
          undefined,
        );
      }
      throw error;
    } finally {
      clearTimeout(timeoutId);
    }
  }
}

File: clients/python/meridian/core/pipeline.py
"""Port of src/core/pipeline.ts — the resilience pipeline that wraps every request.

Order of layers (matching the TypeScript engine):
  auth → rate limiter → retry( circuit breaker( http transport ) ) → normalize → trace
"""

from __future__ import annotations

import time
import uuid
from typing import Any, Awaitable, Callable, Optional

import httpx

from ..adapter import ProviderAdapter
from ..contract import (
    AuthConfig,
    AuthToken,
    BuiltRequest,
    ErrorCategory,
    Headers,
    HttpError,
    MeridianError,
    NormalizedResponse,
    RequestOptions,
    RequestTrace,
    now_utc,
)
from ..strategies.circuit_breaker import ProviderCircuitBreaker
from ..strategies.idempotency import IdempotencyResolver
from ..strategies.rate_limit import RateLimiter
from ..strategies.retry import RetryStrategy
from .endpoint_validator import assert_safe_endpoint
from .sanitizer import sanitize_request_options

Transport = Callable[[BuiltRequest, float], Awaitable["RawResponseT"]]

# Re-export name for readability.
from ..contract import RawResponse as RawResponseT  # noqa: E402


async def _httpx_transport(built: BuiltRequest, timeout: float) -> RawResponseT:
    """Default transport: perform the request with httpx and normalize the result.

    Raises :class:`MeridianError` (network, retryable) on timeout/connection
    failures and :class:`HttpError` on any non-2xx response — mirroring the
    TypeScript ``executeHttpRequest``.
    """
    async with httpx.AsyncClient(timeout=timeout) as client:
        try:
            response = await client.request(
                built.method,
                built.url,
                headers=built.headers,
                content=built.body,
            )
        except httpx.TimeoutException as exc:
            raise MeridianError(
                f"Request timeout after {timeout}s",
                ErrorCategory.NETWORK,
                "",
                True,
                "",
                {"url": built.url, "method": built.method},
            ) from exc
        except httpx.RequestError as exc:
            raise MeridianError(
                "Network request failed. Check your connection and try again.",
                ErrorCategory.NETWORK,
                "",
                True,
                "",
                {"originalError": str(exc)},
            ) from exc

        content_type = response.headers.get("content-type", "")
        try:
            body: Any = response.json() if "application/json" in content_type else response.text
        except ValueError:
            body = {}

        headers = Headers(dict(response.headers))
        if response.status_code >= 400:
            raise HttpError(response.status_code, headers, body)
        return RawResponseT(status=response.status_code, headers=headers, body=body)


class RequestPipeline:
    def __init__(
        self,
        provider: str,
        adapter: ProviderAdapter,
        auth_config: AuthConfig,
        circuit_breaker: ProviderCircuitBreaker,
        rate_limiter: RateLimiter,
        retry_strategy: RetryStrategy,
        idempotency_resolver: IdempotencyResolver,
        base_url: Optional[str] = None,
        timeout: Optional[float] = None,
        auto_generate_idempotency_keys: bool = False,
        transport: Optional[Transport] = None,
        compliance: Optional[dict] = None,
    ) -> None:
        self.provider = provider
        self.adapter = adapter
        self.auth_config = auth_config
        self.circuit_breaker = circuit_breaker
        self.rate_limiter = rate_limiter
        self.retry_strategy = retry_strategy
        self.idempotency_resolver = idempotency_resolver
        self.base_url = base_url
        self.timeout = timeout if timeout is not None else 30.0
        self.auto_generate_idempotency_keys = auto_generate_idempotency_keys
        self.transport = transport or _httpx_transport
        self.compliance = compliance or {}
        self._cached_token: Optional[AuthToken] = None

    async def _get_auth_token(self) -> AuthToken:
        if self._cached_token is not None:
            if self._cached_token.expires_at is None:
                return self._cached_token
            if self._cached_token.expires_at.timestamp() - 60 > time.time():
                return self._cached_token
        token = await self.adapter.auth_strategy(self.auth_config)
        self._cached_token = token
        return token

    async def execute(
        self, endpoint: str, options: Optional[RequestOptions] = None
    ) -> NormalizedResponse:
        options = options or RequestOptions()
        request_id = str(uuid.uuid4())
        method = (options.method or "GET").upper()
        start = time.monotonic()

        assert_safe_endpoint(endpoint, self.provider, request_id)

        if self.auto_generate_idempotency_keys and not options.idempotency_key:
            options.idempotency_key = str(uuid.uuid4())

        idempotency_level = self.idempotency_resolver.get_idempotency_level(method, endpoint)

        # Apply redaction policy (used for observability; the live request keeps
        # the original options). We sanitize to validate the path is exercised.
        sanitize_request_options(
            options,
            pii_redaction=bool(self.compliance.get("pii_redaction")),
            india_mode=bool(self.compliance.get("india_mode")),
        )

        retry_count = {"n": 0}

        async def attempt() -> RawResponseT:
            retry_count["n"] += 1

            async def do_request() -> RawResponseT:
                token = await self._get_auth_token()
                return await self._execute_http(endpoint, options, token)

            return await self.circuit_breaker.execute(do_request)

        try:
            await self._get_auth_token()
            await self.rate_limiter.acquire()

            response = await self.retry_strategy.execute(
                attempt, idempotency_level, bool(options.idempotency_key)
            )

            rate_limit_info = self.adapter.rate_limit_policy(response.headers)
            self.rate_limiter.update_from_headers(rate_limit_info)

            normalized = self.adapter.parse_response(response)
            normalized.meta.request_id = request_id
            duration = (time.monotonic() - start) * 1000
            normalized.meta.trace = RequestTrace(
                retries=retry_count["n"] - 1,
                latency=duration,
                circuit_breaker=self.circuit_breaker.state,
                rate_limit_remaining=rate_limit_info.remaining,
            )
            return normalized
        except Exception as error:  # noqa: BLE001 — mirror the TS catch-all
            meridian_error = self._to_meridian_error(error, request_id)
            if (
                meridian_error.category == ErrorCategory.RATE_LIMIT
                and meridian_error.retry_after is not None
            ):
                seconds = (meridian_error.retry_after - now_utc()).total_seconds()
                self.rate_limiter.handle_429(seconds)
            raise meridian_error from error

    def _to_meridian_error(self, error: Exception, request_id: str) -> MeridianError:
        try:
            parsed = self.adapter.parse_error(error)
        except Exception:  # noqa: BLE001
            parsed = MeridianError(
                str(error), ErrorCategory.PROVIDER, self.provider, False, request_id
            )
        if not parsed.request_id:
            parsed.request_id = request_id
        if not parsed.provider:
            parsed.provider = self.provider
        return parsed

    async def _execute_http(
        self, endpoint: str, options: RequestOptions, auth_token: AuthToken
    ) -> RawResponseT:
        from ..contract import AdapterInput

        built = self.adapter.build_request(
            AdapterInput(
                endpoint=endpoint,
                options=options,
                auth_token=auth_token,
                base_url=self.base_url,
            )
        )
        return await self.transport(built, self.timeout)
